db/db_pebble.go

```diff
@@ -0,0 +1,278 @@
+	}
+	b.db = db
+	return b.TurnOn()
+}
+
+// Stop closes the DB
+func (b *PebbleDB) Stop(_ context.Context) error {
+	if err := b.TurnOff(); err != nil {
+		return err
+	}
+	if err := b.db.Close(); err != nil {
+		return errors.Wrap(ErrIO, err.Error())
+	}
+	return nil
+}
+
+// Get retrieves a record
+func (b *PebbleDB) Get(ns string, key []byte) ([]byte, error) {
+	if !b.IsReady() {
+		return nil, ErrDBNotStarted
+	}
+	v, closer, err := b.db.Get(nsKey(ns, key))
+	if err != nil {
+		if errors.Is(err, pebble.ErrNotFound) {
+			return nil, errors.Wrapf(ErrNotExist, "ns %s key = %x doesn't exist, %s", ns, key, err.Error())
+		}
+		return nil, err
+	}
+	val := make([]byte, len(v))
+	copy(val, v)
+	return val, closer.Close()
+}
+
+// Put inserts a <key, value> record
+func (b *PebbleDB) Put(ns string, key, value []byte) (err error) {
+	if !b.IsReady() {
+		return ErrDBNotStarted
+	}
+	err = b.db.Set(nsKey(ns, key), value, nil)
+	if err != nil {
+		if errors.Is(err, syscall.ENOSPC) {
+			log.L().Fatal("Failed to put db.", zap.Error(err))
+		}
+		err = errors.Wrap(ErrIO, err.Error())
+	}
+	return
+}
+
+// Delete deletes a record,if key is nil,this will delete the whole bucket
+func (b *PebbleDB) Delete(ns string, key []byte) (err error) {
+	if !b.IsReady() {
+		return ErrDBNotStarted
+	}
+	if key == nil {
+		panic("delete whole ns not supported by PebbleDB")
+	}
+	err = b.db.Delete(nsKey(ns, key), nil)
+	if err != nil {
+		if errors.Is(err, syscall.ENOSPC) {
+			log.L().Fatal("Failed to delete db.", zap.Error(err))
+		}
+		err = errors.Wrap(ErrIO, err.Error())
+	}
+	return
+}
+
+// WriteBatch commits a batch
+func (b *PebbleDB) WriteBatch(kvsb batch.KVStoreBatch) error {
+	if !b.IsReady() {
+		return ErrDBNotStarted
+	}
+
+	batch, err := b.dedup(kvsb)
+	if err != nil {
+		return nil
+	}
+	err = batch.Commit(nil)
+	if err != nil {
+		if errors.Is(err, syscall.ENOSPC) {
+			log.L().Fatal("Failed to write batch db.", zap.Error(err))
+		}
+		err = errors.Wrap(ErrIO, err.Error())
+	}
+	return err
+}
+
+func (b *PebbleDB) dedup(kvsb batch.KVStoreBatch) (*pebble.Batch, error) {
+	kvsb.Lock()
+	defer kvsb.Unlock()
+
+	type doubleKey struct {
+		ns  string
+		key string
+	}
+	// remove duplicate keys, only keep the last write for each key
+	var (
+		entryKeySet = make(map[doubleKey]struct{})
+		ch          = b.db.NewBatch()
+	)
+	for i := kvsb.Size() - 1; i >= 0; i-- {
+		write, e := kvsb.Entry(i)
+		if e != nil {
+			return nil, e
+		}
+		// only handle Put and Delete
+		if write.WriteType() != batch.Put && write.WriteType() != batch.Delete {
+			continue
+		}
+		key := write.Key()
+		k := doubleKey{ns: write.Namespace(), key: string(key)}
+		if _, ok := entryKeySet[k]; !ok {
+			entryKeySet[k] = struct{}{}
+			// add into batch
+			if write.WriteType() == batch.Put {
+				ch.Set(nsKey(write.Namespace(), key), write.Value(), nil)
+			} else {
+				ch.Delete(nsKey(write.Namespace(), key), nil)
+			}
+		}
+	}
+	return ch, nil
```

Comment:
better to reverse the order of entries in `ch`, to keep the same order of original batch?

Reply:
I don't think it is necessary, because it doesn't affect the results but adds extra overhead.